### ISSUE TITLE
[Splink4] Use fresh SQLPipeline for all linker methods

### ISSUE THREAD

**Issue by RobinL** [moj-analytical-services/splink#2060]
Rather than a shared mutable [self._pipeline](https://github.com/moj-analytical-services/splink/blob/a6ed0bd2666244fa1ec4345e93c68ee9a003bbee/splink/linker.py#L204), instead use a fresh `pipeline = SQLPipeline()` for all methods such as `predict()`

One challenge with this at the moment is materialisation.

At the moment, we have methods like `_initialise_df_concat_with_tf`, which is difficult to understand due to how it mutates things and returns things:
- It mutates the `self._pipelien` by adding sql to it
- If the user has asked to materialise, it returns a `SplinkDataframe`.  The pipeline will then be flushed because it's been executed
- If not it will have added to the pipeline, and returns `None`

On reason for this behaviour is `_initialise_df_concat_with_tf` is created onced and used many times, so often we want to compute it so it can be cached.

Can this be handled somehow by the pipeline itself?

Specifically, can we extend the concept of the pipeline to allow us to tell it that some parts of the pipeline need to be materialised?

How does this interact with caching?  For any pipeline, can we magically check whether any part of it is already cached?

**Comment by RobinL** [comment]
Probably need to start with a minisplink example using dbapi

**Comment by RobinL** [comment]
Example code:

```
from splink import DuckDBAPI, splink_datasets
from splink.pipeline import SQLPipeline

dbapi = DuckDBAPI()


df = splink_datasets.fake_1000

dbapi.register_table(df, "df")


pipeline = SQLPipeline()
sql = """
select first_name, surname
from df
"""

pipeline.enqueue_sql(sql, "df2")

sql = """
select first_name, count(*) as c
from df2
group by first_name
"""

pipeline.enqueue_sql(sql, "final")

dbapi.sql_pipeline_to_splink_dataframe(pipeline).as_pandas_dataframe()
```

**Comment by RobinL** [comment]
I've tried a solution that modifies sql tasks to allow them to be materialised

```
class SQLTask:
    def __init__(self, sql, output_table_name, materialise=False):
        self.sql = sql
        self.output_table_name = output_table_name
        self.materialise = materialise
```

And then extends SQLPipeline to allow it to have sub-pipelines (be subdivided into CTE sequences) which are executed.  it's possibly viable, but a bit problem is that the pipeline class then needs to somehow be able to predict the name of the output.  Maybe that could be enabled by allowing the a task to know about its `output_table_name` but also a `physical_output_table_name`

But you still have a problem that if you have materialisations, it's not clear what input tables need to carry forward to the next step - i.e. it assumes a single linear dag whereas in general a CTE sequence could be materialised half way, but still need the input tables further down.  

It's still solvable by carrying everything forwards, but i think at that point, we're building a generic dag implementation which is probably a bit too much.

However, iirc materialisation is very rarely required mid pipeline, perhaps only in predict(), so instead, I'm thinking about instead just doing a simple option of:

sql generation methods always do something like:
`self._enqueue_df_concat_with_tf(pipeline)`

And then in the special case that materialisation is needed optionally, we just do
```
if materialise:
   # pipeline.execute
```
or whatever.

But trying to stick with a general pattern of the pipeline simply being used to enqueue stuff and then is executed.  Once executed it should be discarded - in the lnog run, we want to prevent enqueuing once executed

WIll be interesting to observe the extent to which predict() is now decoupled from linker as i go through this

**Comment by RobinL** [comment]
Was closed by #2062
